pkg/health/client/tree.go

```diff
@@ -0,0 +1,192 @@
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+)
+
+const (
+	indentSize              = 3
+	leafMaxWidth            = 40
+	link         decoration = "│"
+	mid          decoration = "├──"
+	end          decoration = "└──"
+)
+
+type decoration string
+
+func newRoot(r string) *node {
+	return &node{val: r}
+}
+
+type node struct {
+	val, meta string
+	parent    *node
+	nodes     []*node
+}
+
+func (n *node) addNode(v string) *node {
+	return n.addNodeWithMeta(v, "")
+}
+
+func (n *node) addNodeWithMeta(v, m string) *node {
+	node := node{
+		parent: n,
+		val:    v,
+		meta:   m,
+	}
+	n.nodes = append(n.nodes, &node)
+
+	return &node
+}
+
+func (n *node) addBranch(v string) *node {
+	return n.addBranchWithMeta(v, "")
+}
+
+func (n *node) addBranchWithMeta(v, m string) *node {
+	b := node{
+		parent: n,
+		meta:   m,
+		val:    v,
+	}
+	n.nodes = append(n.nodes, &b)
+
+	return &b
+}
+
+func (n *node) find(val string) *node {
+	if n.val == val {
+		return n
+	}
+	for _, node := range n.nodes {
+		if node.val == val {
+			return node
+		}
+		if v := node.find(val); v != nil {
+			return v
+		}
+	}
+
+	return nil
+}
+
+func (n *node) asBytes() []byte {
+	w := new(bytes.Buffer)
+	level := 0
+	var levelsEnded []int
+	if n.parent == nil {
+		if n.meta != "" {
+			fmt.Fprintf(w, "%v %v", n.val, n.meta)
+		} else {
+			fmt.Fprintf(w, "%v", n.val)
+		}
+		fmt.Fprintln(w)
+	} else {
+		edge := mid
+		if len(n.nodes) == 0 {
+			edge = end
+			levelsEnded = append(levelsEnded, level)
+		}
+		dumpVals(w, 0, levelsEnded, edge, n)
+	}
+	if len(n.nodes) > 0 {
+		dumpNodes(w, level, levelsEnded, n.nodes)
+	}
+
+	return w.Bytes()
+}
+
+func (n *node) String() string {
+	return string(n.asBytes())
+}
+
+func (n *node) lastNode() *node {
+	c := len(n.nodes)
+	if c == 0 {
+		return nil
+	}
+
+	return n.nodes[c-1]
+}
+
+func dumpNodes(w io.Writer, level int, levelsEnded []int, nodes []*node) {
+	for i, node := range nodes {
+		edge := mid
+		if i == len(nodes)-1 {
+			levelsEnded = append(levelsEnded, level)
+			edge = end
+		}
+		dumpVals(w, level, levelsEnded, edge, node)
+		if len(node.nodes) > 0 {
+			dumpNodes(w, level+1, levelsEnded, node.nodes)
+		}
+	}
+}
+
+func dumpVals(w io.Writer, level int, levelsEnded []int, edge decoration, node *node) {
+	for i := 0; i < level; i++ {
+		if isEnded(levelsEnded, i) {
+			fmt.Fprint(w, strings.Repeat(" ", indentSize+1))
+			continue
+		}
+		fmt.Fprintf(w, "%s%s", link, strings.Repeat(" ", indentSize))
+	}
+
+	val := dumpVal(level, node)
+	if node.meta != "" {
+		c := 4 - level
+		fmt.Fprintf(w, "%s %-"+strconv.Itoa(leafMaxWidth+c*2)+"v%s%v\n", edge, val, strings.Repeat("  ", c), node.meta)
+		return
+	}
+	fmt.Fprintf(w, "%s %v\n", edge, val)
+}
+
+func isEnded(levelsEnded []int, level int) bool {
+	for _, l := range levelsEnded {
+		if l == level {
+			return true
+		}
+	}
+
+	return false
+}
+
+func dumpVal(level int, node *node) string {
+	lines := strings.Split(fmt.Sprintf("%v", node.val), "\n")
```

Comment:
Now that `val` is a string this can be simplified:

```go
lines := strings.Split(node.val, "\n")
```

There are also some other places where `%s` can be used instead of `%v`

Reply:
Great catch! Tx Fabio I'll update.